src/controller/proxy/local.go

```diff
@@ -0,0 +1,181 @@
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/manifest/manifestlist"
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/controller/blob"
+	"github.com/goharbor/harbor/src/core/config"
+	"github.com/goharbor/harbor/src/lib"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/replication/adapter/harbor/base"
+	"github.com/goharbor/harbor/src/replication/model"
+	"io"
+	"sync"
+	"time"
+)
+
+var (
+	mu       sync.Mutex
+	inflight map[string]interface{} = make(map[string]interface{})
+)
+
+// LocalInterface defines operations related to local repo under proxy mode
+type LocalInterface interface {
+	// BlobExist check if the blob exist in local repo
+	BlobExist(ctx context.Context, dig string) (bool, error)
+	// PushBlob push blob to local repo
+	PushBlob(ctx context.Context, p *models.Project, localRepo string, desc distribution.Descriptor, bReader io.ReadCloser) error
+	// PushManifest push manifest to local repo
+	PushManifest(ctx context.Context, p *models.Project, repo string, tag string, mfst distribution.Manifest) error
+	// PushManifestList push manifest list to local repo
+	PushManifestList(ctx context.Context, p *models.Project, repo string, tag string, art lib.ArtifactInfo, man distribution.Manifest) error
+	// CheckDependencies check if the manifest's dependency is ready
+	CheckDependencies(ctx context.Context, man distribution.Manifest, dig string, mediaType string) []distribution.Descriptor
+	// DeleteManifest cleanup delete tag from local cache
+	DeleteManifest(ctx context.Context, repo, ref string)
+}
+
+// local defines operations related to local repo under proxy mode
+type local struct {
+	adapter *base.Adapter
+}
+
+// CreateLocalInterface create the LocalInterface
+func CreateLocalInterface() LocalInterface {
+	return &local{}
+}
+
+// TODO: replace it with head request to local repo
+func (l *local) BlobExist(ctx context.Context, dig string) (bool, error) {
+	return blob.Ctl.Exist(ctx, dig)
+}
+
+func (l *local) init() error {
+	if l.adapter != nil {
+		return nil
+	}
+	registryURL := config.GetCoreURL()
+	reg := &model.Registry{
+		URL: registryURL,
+		Credential: &model.Credential{
+			Type:         model.CredentialTypeSecret,
+			AccessSecret: config.ProxyServiceSecret,
+		},
+	}
+	adapter, err := base.New(reg)
+	l.adapter = adapter
+	return err
+}
+
+func (l *local) PushBlob(ctx context.Context, p *models.Project, localRepo string, desc distribution.Descriptor, bReader io.ReadCloser) error {
+	log.Debugf("Put blob to local registry, localRepo:%v, digest: %v", localRepo, desc.Digest)
+	if err := l.init(); err != nil {
```

Comment:
Why we need to call `init` for every function call?

Reply:
+1 Why not move the code of init into Create.